Source/Particles/Collision/BackgroundStopping/BackgroundStopping.cpp

```diff
@@ -0,0 +1,269 @@
+                doBackgroundStoppingOnIonsWithinTile(pti, dt, cur_time, species_mass, species_charge);
+            }
+
+            if (cost && WarpX::load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::Timers)
+            {
+                amrex::Gpu::synchronize();
+                wt = amrex::second() - wt;
+                amrex::HostDevice::Atomic::Add(&(*cost)[pti.index()], wt);
+            }
+        }
+
+    }
+}
+
+void BackgroundStopping::doBackgroundStoppingOnElectronsWithinTile (WarpXParIter& pti, amrex::Real dt, amrex::Real t,
+                                                                    amrex::Real species_mass, amrex::Real species_charge)
+{
+    using namespace amrex::literals;
+
+    // So that CUDA code gets its intrinsic, not the host-only C++ library version
+    using std::sqrt, std::abs, std::log, std::exp;
+
+    // get particle count
+    const long np = pti.numParticles();
+
+    // get background particle mass
+    amrex::Real mass_e = m_background_mass;
+
+    // setup parsers for the background density and temperature
+    auto n_e_func = m_background_density_func;
+    auto T_e_func = m_background_temperature_func;
+
+    // get Struct-Of-Array particle data, also called attribs
+    auto& attribs = pti.GetAttribs();
+    amrex::ParticleReal* const AMREX_RESTRICT ux = attribs[PIdx::ux].dataPtr();
+    amrex::ParticleReal* const AMREX_RESTRICT uy = attribs[PIdx::uy].dataPtr();
+    amrex::ParticleReal* const AMREX_RESTRICT uz = attribs[PIdx::uz].dataPtr();
+
+    // May be needed to evaluate the density and/or temperature functions
+    auto GetPosition = GetParticlePosition(pti);
+
+    amrex::ParallelFor(np,
+        [=] AMREX_GPU_HOST_DEVICE (long ip)
+        {
+
+            amrex::ParticleReal x, y, z;
+            GetPosition.AsStored(ip, x, y, z);
+            amrex::Real const n_e = n_e_func(x, y, z, t);
+            amrex::Real const T_e = T_e_func(x, y, z, t)*PhysConst::kb;
+
+            // This implements the equation 14.12 from Introduction to Plasma Physics,
+            // Goldston and Rutherford, the slowing down of beam ions due to collisions with electrons.
+            // The equation is written as dV/dt = -alpha*V, and integrated to
+            // give V(t+dt) = V(t)*exp(-alpha*dt)
+
+            amrex::Real constexpr pi = MathConst::pi;
+            amrex::Real constexpr ep0 = PhysConst::ep0;
+            amrex::Real constexpr q_e = PhysConst::q_e;
+            amrex::Real constexpr q_e2 = q_e*q_e;
+            amrex::Real constexpr ep02 = ep0*ep0;
+
+            amrex::Real const Zb = abs(species_charge/q_e);
+
+            amrex::Real const vth = sqrt(3._rt*T_e/mass_e);
+            amrex::Real const wp = sqrt(n_e*q_e2/(ep0*mass_e));
+            amrex::Real const lambdadb = vth/wp;
+            amrex::Real const lambdadb3 = lambdadb*lambdadb*lambdadb;
+            amrex::Real const loglambda = log((12._rt*pi/Zb)*(n_e*lambdadb3));
```

Comment:
@dpgrote as discussed via email:
we need to add extra care here that the argument to `log` are >0. Especialliy:
- `vth` might be zero
- `n_e` migth be zero
- `Zb`: the species might be a neutral

In Andreas Kemp's report, the issue is with `vth` not being `>0`.

Reply:
Addressed in #3163